### ISSUE TITLE
wyng requirements (qubes vs qubes-ssh mode, remote storage providers for RW/RO subaccount... and fragility of SSHFS.

### ISSUE THREAD

**Issue by tlaurion** [tasket/wyng-backup#83]
Might be totally out of scope of this project since my aim is to be able to store/retrieve OEM disk images LVM content, not private content (think updated TemplateVMs, AppVMs and dom0, with possibility to revert to known signed state when needed).

As you know, finding a good hoster with proper price for remote data storage,  proper ownership/permission management of data storage (multi-user support, permitting RO access of archived content) is not so easy to find.

A while ago, I started a PoC with Rsync.net, having them setup a two user subaccounts setup. 

- RW, main user (OEM qubes to deploy) can write and own all files in its home directory and RO account directory and subdirectories (wyng-rsync_net-RW qube used to deploy wyng backups in OEM mode, with public key matching RW accesses)
- RO subuser account has its home directory under RW home directory, which can read the content of it's own directories but cannot change any of the files, owned by the RW account. (wyng-rsync_net qube aimed to be deployed on OEM image to verify/restore QubesOS into OEM known state).

All good theoretically.

Some clarifications needed on wyng requirements (qubes-ssh mode): 
- `mountpoint -q` is a requirement from permitted commands to be ran over remote ssh server right? Not from qubes-ssh host. [Can that be mitigated](https://www.rsync.net/resources/howto/remote_commands.html)? rsync.net use BSD and as you can see, they offer access to a really limited subset of remote commands.
- python helper-script code (and python requirements) are to be ran on the qube that connects to the remote ssh server, right? Not the remote ssh server over ssh?

If `python` and `mountpoint -q` are to be ran on the remote ssh server from `quebes-ssh` mode, that leaves the user to use SSHFS to mount the remote ssh location locally on  wyng-qubes AppVM in `qubes://` mode, not `qubes-ssh://` mode. 

SSHFS is damn slow and seems to go over its own subset of problems (caching, listing, access of files, linking unlinking is overwhelmingly slow where maybe unusable)....

If you could shed  some light on the above two points. I would wait on those answers prior of troubleshooting the `qubes://` mode over SSHFS which is causing me issues right now. SSHFS has a really high performance cost and seems a lot less reliable altogether.... With what seems to be really [limited support of hardlinks from their usage of FUSE API](https://github.com/libfuse/sshfs/issues/109). I was not able to complete a single send, even less test dedup, with different errors happening even if the storage is still mounted properly and alive, with proper caching options at mount and mountpoint being kept alive. But that will be digged down later on, where tweaking can still happen.

But if SSHFS is not really usable, that would lead into finding other hosters willing to have multiple subaccounts linked to same hosting plan, where rsync.net offers pretty unbeatable and stable prices based on used storage space, not bandwidth usage..

Another option to be investigated would be to have a locally `qubes://` mode mounted loopback device over SSHFS under appvm. Intuition here is that the IO and hardlink problems would probably vanish at the cost of added complexity, while the costs of hosting would not depend of current used storage, which would be sad because of dedup and pruning possibilites. SSHFS usage seems to be an option to be ditched altogether unless a LUKS encrypted loopback device is needed (not in my case) and even less when encryption support is complete for cloud storage. 

Basically, my question is about clarifying more clearly wyng requirements on `qubes://` and `qubes-ssh://` modes, and hoster's  binaries access requirement on ssh server when `qubes-ssh://` mode is envisaged. Thanks and as always, sorry for the number of words in this issue.

**Comment by tlaurion** [comment]
@tasket 

Maybe documenting [where those binaries](https://github.com/tasket/wyng-backup/blob/cb04df4f72218d65313e7a5626fb2925ea415a2e/wyng#L387-L397) are to be found on which side (Host to be backuped, qubes, ssh?)

**Comment by tasket** [comment]
The mountpoint cmd is run on the remote. Possibly you could override it by creating a symlink to `/usr/bin/true` or a shell script that does `exit 0`.

Cmd calls on remote don't (yet) follow strict adherence to the CP paths; those were originally intended for local cmds. I don't intend to use the absolute CP path for mountpoint, so that leaves the possibility of placing the symlink or script in a location that comes earlier in `$PATH`.

`dest_helper.py`: Sorry, no. Its not targeted at 'local connection qube'. Its targeted at _'OS that uses the destination filesystem'_ which for qubes-ssh: means the remote host on the far end of the ssh connection.

As for what remote system cmds it needs: `python3, sync, rm, mv, makedir, cat & tar`. I think `awk` is currently only run locally, so not a concern.

Yes `sshfs` along with the qubes: (not qubes-ssh:) specifier can get you around these hurdles. I think sshfs can be pretty slow when used with the kind of LUKS+Ext4 loopback container (in dom0) suggested in the Readme, and this is how I use it. So I'm sorry to hear its also slow without the container layer. I noticed there is a lot of [sshfs tuning](https://www.admin-magazine.com/HPC/Articles/Sharing-Data-with-SSHFS) advice out there; a 100% speed improvement seems pretty good.

I do not think doing loopback in a qube would be faster; using any loopback over remote means that internal filesystem 'chatter' goes over the net and its noticeable.

Also consider that NFS or Samba might offer a performance advantage.

---

Bottom line is that `mountpoint` can be worked around one way or another for qubes-ssh: mode, but running Python on the remote is _currently_ non-negotiable.

**Comment by tlaurion** [comment]
@tasket 

So, PoC is going its way over rsync.net

1- Original backup was pushed with rsync from local wyng-backups-vm to rsync.net, with wyng-backups-vm private volume LVM attached to rsync-RW account to do initial rsync.
2- wyng-rsync-RW appvm was used to push delta with dedup, which worked over SSHFS (FS mounted with ssh RW account, so can write in directory that is then mounted by the wyng-rsync-RO AppVM under same mountpoint which is the same local dir on both AppVMs)
3- wyng-rsync-RO appvm is now used to do an wyng arch-init from=qubes://wyng-rsync-RO with proper subdir (same dir is mounted in both RO and RW AppVM)

It seems that RW is needed to do `arch-init from`. That is problematic if that is the case, the objective here being that send operations are denied over RO AppVM connected to SSHFS over mountpoint that cannot write anything... Am I misunderstanding something here?

**Comment by tasket** [comment]
@tlaurion Write access is not a strict operational requirement (for receive, list, etc.). But the online status checker doesn't discern between rw and ro commands. I'll look into changing that.

**Comment by tlaurion** [comment]
@tasket keep me posted. PoC will stop there for online receive tests until you tag me back!
I reviewed configs and directories/files permissions on remote storage and they seem fine. 
If you want me to test anything ,as always, let me know.

**Comment by tasket** [comment]
@tlaurion See if this works.

**Comment by tlaurion** [comment]
> @tlaurion See if this works.
@tasket  when doing arch-init from RO qubes having RO access to SSHFS mountpoint:
```
Traceback
2928 get_config_remote()
765 get_config_remote
line 1151 do_exec
subprocess.CalledProcessErrror: Commande /bin/tar -xfz - returned non-zero exit status 2
```

**Comment by tasket** [comment]
@tlaurion What is the command line?

**Comment by tlaurion** [comment]
@tasket 
`sudo wyng --meta-dir-/var/lib/wyng-rsync_net-Insurgo --from=qubes://wyng-rsync_net-Insurgo/ --subdir=home/user/rsync-net_mountpoint arch-init --debug`
```

Where:
qvm-run --no-gui --pass-io wyng-rsync_net-Insurgo "ls /home/user/rsync-net_mountpoint"
[...]
dxwxr-xr-x  11971 18407 8 Jul 13:59 wyng.backup
``` 
Where 
- 11971 is RW account
- 18407 is RO account (used here to mount)

And where:
- `qvm-run --no-gui --pass-io wyng-rsync_net-Insurgo "cat /home/user/rsync-net_mountpoint/wyngbackup/default/archive.ini` successfully shows content.
- `qvm-run --no-gui --pass-io wyng-rsync_net-Insurgo "touch /home/user/rsync-net_mountpoint/test` fails with "Permission denied"

**Comment by tasket** [comment]
@tlaurion What is in the /tmp/wyng/err.log? The update I just posted allows for more log detail if `tar` reports an error.

**Comment by tlaurion** [comment]
@tasket issue with last volume pushed without ashes mount containing idmap. 

Investigating

**Comment by tlaurion** [comment]
So ssh remote needed commands are:

> python3, sync, rm, mv, makedir, cat & tar. I think awk is currently only run locally, so not a concern.

> mountpoint can be worked around one way or another for qubes-ssh: mode, but running Python on the remote is currently non-negotiable.

**Comment by tasket** [comment]
@tlaurion Yes, unless you use something like sshfs (with Wyng connection type `internal` or `qubes`) in which case you would probably not need any of them.

**Comment by tasket** [comment]
@tlaurion IIRC you had performance problems accessing your destination via sshfs, possibly because of limited directory caching.  Also just noticed this significant issue:
>    Hardlinks will always appear as individual files when seen through an SSHFS  mount,
       i.e. they will appear to have different inodes and an st_nlink value of 1.

This might not effect Wyng `send -d` at all, but it could have a big slowdown effect on `arch-deduplicate`.  Other functions probably not affected.

---


I just noticed this –a FUSE filesystem for rsync–
https://github.com/zaddach/fuse-rsync

It might be worth a try.

**Comment by tlaurion** [comment]
@tasket from my not so extended research for hosting, providers with ssh access will not be so happy providing python....

**Comment by tlaurion** [comment]
And yes. sshfs was really impacting performance to a point that it was unusable. Mounting a loopfs from qubes+ssh mitigated this, but locking for sync was also a disaster.

**Comment by tasket** [comment]
It really should be do-able and efficient without using `losetup`, just using a FUSE filesystem.  I opened issue #101 about sshfs performance.

**Comment by tasket** [comment]
Re: Python availability...

This is of course for the Wyng helper script, which is written in python.  Unlike the main Wyng program, it isn't tied to anything especially pythonic.  But some type of reasonably quick interpreter is needed and I think there are some decent python replacement candidates from `perl` and `tcl` to rudimentary ones like `awk`.  The question is, if a hosting provider doesn't like to offer python, wouldn't they also tend to restrict those other candidates as well?

**Comment by tlaurion** [comment]
> @tlaurion IIRC you had performance problems accessing your destination via sshfs, possibly because of limited directory caching. 

Just realized I posted in wrong issue https://github.com/tasket/wyng-backup/issues/101#issuecomment-1164618307
`Compression=no,kernel_cache,cache=yes` was in last test-bench


>Also just noticed this significant issue:
> 
> > Hardlinks will always appear as individual files when seen through an SSHFS  mount,
> > i.e. they will appear to have different inodes and an st_nlink value of 1.

This is really problematic :(

> 
> This might not effect Wyng `send -d` at all, but it could have a big slowdown effect on `arch-deduplicate`. Other functions probably not affected.
> 
> I just noticed this –a FUSE filesystem for rsync– https://github.com/zaddach/fuse-rsync
> 
> It might be worth a try.

Will look, but as of now, without a remote server, local tests are amazing but have less provable scope. Having recommended servers, or proper implementations on top of aws or other seems to be the next step making this reach wyng-backup proper user base.

Would love to distribute OEM revertible states to customer base, but as of now, this is not possible outside of the static state provided through wyng-backup qube! :)
